xtask/src/task_phd.rs

```diff
@@ -0,0 +1,349 @@
+                    $(
+                        $arg => {
+                            let val = arg_iter.next().ok_or_else(|| {
+                                anyhow::anyhow!("Missing value for argument `{}`", $arg)
+                            })?;
+                            cargo_log!("Overridden", "{} {val:?}", $arg);
+                            $var = Some(val);
+                        }
+                    ),+
+                    _ => bonus_args.push(arg),
+                }
+            }
+        }
+        args! {
+            args::PROPOLIS_BASE => propolis_base_branch,
+            args::PROPOLIS_CMD => propolis_local_path,
+            args::CRUCIBLE_COMMIT => crucible_commit,
+            args::ARTIFACTS_TOML => artifacts_toml,
+        }
+    }
+
+    let propolis_local_path = match propolis_local_path {
+        Some(path) => path.into(),
+        None => {
+            let bin = build_bin("propolis-server")?;
+            let path = bin
+                .path()
+                .try_into()
+                .context("Propolis server path is not UTF-8")?;
+            relativize(path).to_path_buf()
+        }
+    };
+
+    let meta = cargo_metadata::MetadataCommand::new()
+        .no_deps()
+        .exec()
+        .context("Failed to run cargo metadata")?;
+    let phd_dir = relativize(&meta.target_directory).join("phd");
+    let artifact_dir = phd_dir.join("artifacts");
+    mkdir(&artifact_dir, "artifact directory")?;
+
+    let tmp_dir = {
+        let mut tmp_dir = phd_dir.join("tmp");
+        let now = time::SystemTime::now();
+        delete_old_tmps(&tmp_dir, now)?;
+        tmp_dir.push(
+            now.duration_since(time::UNIX_EPOCH).unwrap().as_secs().to_string(),
+        );
+        tmp_dir
+    };
+
+    mkdir(&tmp_dir, "temp directory")?;
+
+    let artifacts_toml =
+        artifacts_toml.map(Utf8PathBuf::from).unwrap_or_else(|| {
+            // if there's no explicitly overridden `artifacts.toml` path,
+            // determine the default one from the workspace path.
+            relativize(&meta.workspace_root)
+                .join("phd-tests")
+                .join("artifacts.toml")
+        });
+
+    if artifacts_toml.exists() {
+        cargo_log!("Found", "artifacts.toml at `{artifacts_toml}`")
+    } else {
+        anyhow::bail!("Missing artifacts config `{artifacts_toml}`!");
+    }
+
+    let status = run_exit_code(
+        phd_runner
+            .command()
+            .arg(args::RUN)
+            .arg(args::PROPOLIS_CMD)
+            .arg(&propolis_local_path)
+            .arg(args::CRUCIBLE_COMMIT)
+            .arg(crucible_commit.unwrap_or("auto"))
+            .arg(args::PROPOLIS_BASE)
+            .arg(propolis_base_branch.unwrap_or("master"))
+            .arg(args::ARTIFACTS_TOML)
+            .arg(&artifacts_toml)
+            .arg("--artifact-directory")
+            .arg(&artifact_dir)
+            .arg("--tmp-directory")
+            .arg(&tmp_dir)
```

Comment:
I'd probably make at least the artifact directory overrideable. In the future when we specify more and larger artifacts in the checked-in TOML, I think it'd be valuable to be able to elect to stash those in a directory that isn't subject to being pruned by `cargo clean`.

(I suppose the tmp directory could be overridden too, but I don't think the argument for it is as strong, outside maybe of uniformity with the other switches.)

Reply:
oh, i thought i'd done this, looks like maybe i forgot to push that release!

Reply:
alright, as of 6cdbbafd617d05c704f1eed1ed3101879d1a0523, you can now do:
```console
$ cargo xtask phd run --artifact-directory ~/.cache/phd/artifacts
   Compiling phd-runner
    Finished phd-runner dev [unoptimized + debuginfo] in 0.23s
  Overridden --artifact-directory "/home/eliza/.cache/phd/artifacts"
   Compiling propolis-server
    Finished propolis-server dev [unoptimized + debuginfo] in 51.95s
     Created artifact directory `/home/eliza/.cache/phd/artifacts`
     Created temp directory `target/phd/tmp/1706309041`
       Found artifacts.toml at `phd-tests/artifacts.toml`
     Running target/x86_64-unknown-linux-gnu/debug/phd-runner run \
        --propolis-server-cmd target/x86_64-unknown-linux-gnu/debug/propolis-server \
        --crucible-downstairs-commit auto \
        --base-propolis-branch master \
        --artifact-toml-path phd-tests/artifacts.toml \
        --artifact-directory /home/eliza/.cache/phd/artifacts \
        --tmp-directory target/phd/tmp/1706309041
2024-01-26T22:44:01.294551Z  WARN phd_runner: 27: Failed to enable one or more kernel options, some tests may not work error=set_vmm_globals requires illumos
```